_posts/2022-04-22-switch-to-java-11.adoc

```diff
@@ -0,0 +1,138 @@
+Testcontainers allow starting and stopping containers programmatically when needed, 
+allows defining the images directly in code so we don't need any additional `Dockerfile`
+, and cleaning up the container is just an implementation detail hidden in the test itself.
+Having the ability to run the tests directly from the IDE is a nice benefit on top of these things.
+
+The only tricky thing when using testcontainers was that when we tried to remove the files using Docker `cmd` command,
+testcontainers randomly failed, stating that the container didn't start in spite of the fact that all Cassandra files were actually deleted.
+The container probably ran so fast that it finished before testcontainers noticed it.
+Finally, we solved it by adding a short `sleep` in the container and executing an additional command in the container which does the cleanup.
+
+The final cleanup code using testcontainers looks like this:
+
+[source, xml]
+----
+@AfterClass
+public static void tearDownClass() throws IOException, InterruptedException {
+    destroyTestKeyspace();
+    cassandra.stop();
+
+    GenericContainer clenaup = new GenericContainer(new ImageFromDockerfile()
+            .withDockerfileFromBuilder(builder -> builder
+                    .from("eclipse-temurin:8-jre-focal")
+                    .volume("/var/lib/cassandra")
+                    .cmd("sleep", "10") // Give testcontainers some time to find out container is running.
+                    .build()))
+            .withFileSystemBind(cassandraDir, CASSANDRA_SERVER_DIR, BindMode.READ_WRITE);
+    clenaup.start();
+    clenaup.execInContainer(
+            "rm", "-rf",
+            CASSANDRA_SERVER_DIR + "/data",
+            CASSANDRA_SERVER_DIR + "/cdc_raw_directory",
+            CASSANDRA_SERVER_DIR + "/commitlog",
+            CASSANDRA_SERVER_DIR + "/hints",
+            CASSANDRA_SERVER_DIR + "/saved_caches");
+    clenaup.stop();
+}
+----
+
+Once we solved the issue with Cassandra tests, we were mostly done and were ready to use Java 11/17.
+
+== Open issues
+
+We need more battle testing to be sure that everything works well with Java 11/17.
+Your help with testing and bug reports would be very valuable here and more than welcome.
+Currently we are aware of one minor unsolved issue related to the Java update.
+Some IDEs cannot distinguish between `maven.compiler.release` and `maven.compiler.testRelease` (or it's not very clear to us who to set it up).
+For example this test using https://openjdk.java.net/jeps/378[text block] is marked as an error in the IDE:
+
+[.centered-image.responsive-image]
+====
+++++
+<img src="/assets/images/2022-04-22-switch-to-java-11/idea_error.png" style="max-width:90%;" class="responsive-image">
+++++
+Test using text block in Idea IDE.
+====
+
+You can manually set Java level to 17, but in this case you may unintentionally use Java > 11 features in non-test code without IDE letting you know.
```

Comment:
```suggestion
You can manually set Java level to 17, but in this case you may unintentionally use Java > 11 features in non-test code without the IDE letting you know.
```

Worth pointing out that this is a nuisance but not a show stopper, as the next Maven build (latest on CI) would fail in that case.